theano/compile/builders.py

```diff
@@ -115,40 +64,88 @@ def __hash__(self):
         # TODO: use internal variables in hash
         return hash(type(self))
 
+    def grad(self, inputs, output_grads):
+        if self.cached_grad_ops:
+            return self.grad_ops(inputs, output_grads)
+
+        grad_inps = self.internal_inputs + output_grads
+        upstream_grads = dict(izip(self.internal_outputs, output_grads))
+        if self.grad_ops is not None:
+            grad_ops_l = self.grad_ops
+            if isinstance(grad_ops_l, list):
+                assert len(grad_ops_l) <= len(self.internal_inputs)
+                if len(grad_ops_l) < len(self.internal_inputs):
+                    grad_ops_l += [None] * (
+                        len(self.internal_inputs) - len(grad_ops_l))
+                # It is normal if some inputs are not needed in order
+                # to compute the gradient, so we ignore them.
+                gs = [go if go else type(self)(
+                    grad_inps,
+                    (lambda g: g if g else (lambda *a:None))(
+                        theano.gradient.grad(
+                            cost=None,
+                            known_grads=upstream_grads,
+                            wrt=[inp],
+                            disconnected_inputs='ignore')
+                    ), on_unused_input='ignore'
+                ) for go, inp in izip(grad_ops_l, self.internal_inputs)]
```

Comment:
But then the base class needs to know whether it needs to create those `type(self)` instances or not, at which point it may really be easier to just add an `inline` argument (or a `precompile` argument with inverse semantics) to `OpFromGraph` instead of adding additional classes. What do you think?